firmware/library/L2_HAL/sensors/optical/apds_9960.hpp

```diff
@@ -353,98 +335,101 @@ class Apds9960 : public Apds9960Interface
-                 down_sensitivity_)  // Down
+        // Check which photodiode pair sensed more IR light
+        if (abs((gfifo_data_[i] - gfifo_data_[i + 1])) >
+            abs((gfifo_data_[i + 2] - gfifo_data_[i + 3])))
         {
-          points_[index_] = Direction::kDown;
-          index_++;
-        }
-      }
-      else
-      {
-        // If a photodiode has a smaller value,
-        // then that photodiode was covered and movement detected
-        if ((gfifo_data_[i + 2] - gfifo_data_[i + 3]) <
-            left_sensitivity_)  // Left
-        {
-          points_[index_] = Direction::kLeft;
-          index_++;
+          // If a photodiode has a smaller value,
+          // then that photodiode was covered and movement detected
+          if ((gfifo_data_[i] - gfifo_data_[i + 1]) < kUpSensitivity)  // Up
+          {
+            points.push_back(Direction::kUp);
+          }
+          else if ((gfifo_data_[i] - gfifo_data_[i + 1]) >
+                   kDownSensitivity)  // Down
+          {
+            points.push_back(Direction::kDown);
+          }
         }
-        else if ((gfifo_data_[i + 2] - gfifo_data_[i + 3]) >
-                 right_sensitivity_)  // Right
+        else
         {
-          points_[index_] = Direction::kRight;
-          index_++;
+          // If a photodiode has a smaller value,
+          // then that photodiode was covered and movement detected
+          if ((gfifo_data_[i + 2] - gfifo_data_[i + 3]) <
+              kLeftSensitivity)  // Left
+          {
+            points.push_back(Direction::kLeft);
+          }
+          else if ((gfifo_data_[i + 2] - gfifo_data_[i + 3]) >
+                   kRightSensitivity)  // Right
+          {
+            points.push_back(Direction::kRight);
+          }
         }
       }
       // Check the sum of all photodiodes
       if ((gfifo_data_[i] + gfifo_data_[i + 1] + gfifo_data_[i + 2] +
-           gfifo_data_[i + 3]) > near_sensitivity_)  // Near
+           gfifo_data_[i + 3]) > kNearSensitivity)  // Near
       {
         near_count_++;
       }
       else if ((gfifo_data_[i] + gfifo_data_[i + 1] + gfifo_data_[i + 2] +
-                gfifo_data_[i + 3]) < far_sensitivity_)  // Far
+                gfifo_data_[i + 3]) < kFarSensitivity)  // Far
       {
         far_count_++;
       }
     }
   }
-  Gesture DecodeGestureData(uint8_t level) override
+  Gesture DecodeGestureData() override
   {
     Gesture result = kError;
+    bool found     = false;
 
-    if (level > 15)
+    if (far_count_ > 5)
     {
-      if (far_count_ > 5)
-      {
-        result = kFAR;
-      }
-      else if (near_count_ > 5)
-      {
-        result = kNEAR;
-      }
+      result = kFAR;
+      found  = true;
+    }
+    else if (near_count_ > 5)
+    {
+      result = kNEAR;
+      found  = true;
     }
 
-    for (int i = 0; i < kMaxDataSize; i++)
+    for (uint8_t i = 0; i < points.size() && !found; i++)
```

Comment:
```suggestion
    for (uint8_t i = 0; i < points.size() - 1 && !found; i++)
```
To keep you from access beyond the array via `points[i + 1]`